server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java

```diff
@@ -0,0 +1,598 @@
+                                targetNode.nodeId()
+                            );
+                        }
+                        routingNodes.relocateShard(
+                            shard,
+                            targetNode.nodeId(),
+                            allocation.clusterInfo().getShardSize(shard, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE),
+                            allocation.changes()
+                        );
+                        eligibleNodes.offer(targetNode);
+                        break;
+                    } else {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace(
+                                "Cannot move shard: {} to node: [{}]. Decisions: [{}]",
+                                shardShortSummary(shard),
+                                targetNode.nodeId(),
+                                currentShardDecision.getDecisions()
+                            );
+                        }
+
+                        Decision nodeLevelDecision = allocation.deciders().canAllocateAnyShardToNode(targetNode, allocation);
+                        if (nodeLevelDecision.type() == Decision.Type.YES) {
+                            logger.debug("Node: [{}] can still accept shards. Adding it back to the queue.", targetNode.nodeId());
+                            eligibleNodes.offer(targetNode);
+                            nodesCheckedForShard.add(targetNode.nodeId());
+                        } else {
+                            logger.debug("Node: [{}] cannot accept any more shards. Removing it from queue.", targetNode.nodeId());
+                        }
+
+                        // Break out if all nodes in the queue have been checked for this shard
+                        if (eligibleNodes.stream().allMatch(rn -> nodesCheckedForShard.contains(rn.nodeId()))) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Performs heuristic, naive weight-based balancing for remote shards within the cluster by using average nodes per
+     * cluster as the metric for shard distribution.
+     * It does so without accounting for the local shards located on any nodes within the cluster.
+     */
+    @Override
+    void balance() {
+        Iterator<RoutingNode> nodesIter = routingNodes.iterator();
+        List<RoutingNode> nodeList = new ArrayList<>();
+        while (nodesIter.hasNext()) {
+            RoutingNode rNode = nodesIter.next();
+            nodeList.add(rNode);
+        }
+        logger.trace("Performing balancing for remote shards.");
+
+        if (nodeList.isEmpty()) {
+            logger.info("No eligible remote nodes found to perform balancing");
+            return;
+        }
+
+        ObjectIntHashMap<String> primaryShardCount = new ObjectIntHashMap<>();
+        int totalPrimaryShard = 0;
+        for (RoutingNode node : nodeList) {
```

Comment:
Can we extract this logic to a different method?

Reply:
Done.